### ISSUE TITLE
MultivariateNormal backprop performance issue related to broadcasting

### ISSUE THREAD

**Issue by fatvlady** [pytorch/pytorch#43837]
## 🐛 Bug

Currently multivariate normal broadcasts `loc` and `covariance_matrix` tensors together, effectively creating `loc_` tensor of shape `(*batch_dims, n, n)` and slicing from it after instead of creating tensor of shape `(*batch_dims, n)`. Although there's not much performance difference during forward pass, it creates massive `fill_` operation during backward pass which just zeroes gradient in `loc_` gradient tensor as you can notice in profiler trace. Moreover, it probably sums gradients over the whole `loc_` grad tensor after, but this is not that pronounced in profiler trace. In the end, both fill and sum ops are useless.

## To Reproduce
In this script I tried to provide possible fix, although I do not like double `broadcast_tensor` in each branch. Maybe there's a way to do it in more efficient and elegant way. I was looking for `broadcast_to` method which favours AD flow, but found only #8076.
```
import torch
import torch.distributions as td
from torch.autograd import profiler, grad

torch.set_default_dtype(torch.float64)
torch.set_num_threads(1)


class StubMultivariateNormal(td.MultivariateNormal):
    def __init__(self, loc, covariance_matrix=None, precision_matrix=None, scale_tril=None, validate_args=None):
        super().__init__(loc, covariance_matrix, precision_matrix, scale_tril, validate_args)
        # retry procedure with broadcasting
        loc_ = loc.unsqueeze(-1)  # temporarily add dim on right
        if scale_tril is not None:
            if scale_tril.dim() < 2:
                raise ValueError("scale_tril matrix must be at least two-dimensional, "
                                 "with optional leading batch dimensions")
            self.scale_tril, _ = torch.broadcast_tensors(scale_tril, loc_)
            self.loc, _ = torch.broadcast_tensors(loc, scale_tril[..., 0])
        elif covariance_matrix is not None:
            if covariance_matrix.dim() < 2:
                raise ValueError("covariance_matrix must be at least two-dimensional, "
                                 "with optional leading batch dimensions")
            self.covariance_matrix, _ = torch.broadcast_tensors(covariance_matrix, loc_)
            self.loc, _ = torch.broadcast_tensors(loc, covariance_matrix[..., 0])
        else:
            if precision_matrix.dim() < 2:
                raise ValueError("precision_matrix must be at least two-dimensional, "
                                 "with optional leading batch dimensions")
            self.precision_matrix, _ = torch.broadcast_tensors(precision_matrix, loc_)
            self.loc, _ = torch.broadcast_tensors(loc, precision_matrix[..., 0])


if __name__ == '__main__':
    torch.random.manual_seed(1234)
    loc = torch.randn(1000, 800)
    target = torch.randn(1000, 800)
    cov = target.transpose(-1, -2) @ target / 1000
    loc.requires_grad_()
    dist1 = td.MultivariateNormal(loc=loc, covariance_matrix=cov)
    dist2 = StubMultivariateNormal(loc=loc, covariance_matrix=cov)
    
    # run with modified distribution first to ensure that warmup does not swindle the benchmark
    with profiler.profile() as prof:
        with profiler.record_function('dist2_grad'):
            grad(dist2.log_prob(target).sum(), loc)
    print(prof.key_averages().table(sort_by='cpu_time_total', row_limit=5))
    
    # run with torch's MultivariateNormal
    with profiler.profile() as prof:
        with profiler.record_function('dist1_grad'):
            grad(dist1.log_prob(target).sum(), loc)
    print(prof.key_averages().table(sort_by='cpu_time_total', row_limit=5))

```
### Output
```
------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                            Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  
------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
dist2_grad                      11.04%           5.343ms          100.00%          48.410ms         48.410ms         1                
triangular_solve                0.11%            54.597us         58.66%           28.396ms         14.198ms         2                
_triangular_solve_helper        44.83%           21.704ms         58.52%           28.331ms         14.166ms         2                
TriangularSolveBackward         0.05%            23.087us         29.88%           14.463ms         14.463ms         1                
copy_                           15.76%           7.629ms          15.76%           7.629ms          1.271ms          6                
------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 48.410ms

------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                            Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls  
------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
dist1_grad                      6.92%            116.405ms        100.00%          1.681s           1.681s           1                
SelectBackward                  0.00%            13.233us         77.98%           1.311s           1.311s           1                
fill_                           77.64%           1.305s           77.64%           1.305s           186.446ms        7                
zeros                           0.00%            6.963us          77.62%           1.305s           652.349ms        2                
zero_                           0.00%            9.784us          77.61%           1.305s           652.337ms        2                
------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 1.681s
```

## Environment
```
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Fedora 32 (Workstation Edition) (x86_64)
GCC version: (GCC) 10.2.1 20200723 (Red Hat 10.2.1-1)
Clang version: 10.0.0 (Fedora 10.0.0-2.fc32)
CMake version: version 3.17.4

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2070 SUPER
Nvidia driver version: 450.66
cuDNN version: /usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip3] numpy==1.17.4
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[pip3] torchviz==0.0.1
[conda] Could not collect
```

cc @ezyang @SsnL @albanD @zou3519 @gqchen @VitalyFedyunin @ngimel

**Comment by fatvlady** [comment]
@ezyang @gchanan @fritzo

**Comment by fritzo** [comment]
It seems like a good solution would be to implement a fast jittable helper like [np.broadcast(...).shape](https://numpy.org/devdocs/reference/generated/numpy.broadcast.html) or Pyro's [broadcast_shape(...)](https://github.com/pyro-ppl/pyro/blob/7c2c22c10dffda8a33ffbd593cc8d58819959e40/pyro/distributions/util.py#L151). In Pyro we use `broadcast_shape()` extensively for multivariate distribution reshaping. Does anyone see performance reasons to avoid a pattern like the following?
```py
shape = torch.broadcast_shape(loc.shape, scale_tril.shape[:-1])
loc = loc.expand(shape)
scale_tril = scale_tril.expand(shape + (-1,))
```

cc @fehiepsi who implemented the `._unbroadcasted_*` speedups.

**Comment by fehiepsi** [comment]
Yes, I agree that the `expand` logic here can be expensive.

@fritzo I think it is better to do `self.scale_tril = scale_tril.expand(shape + (-1,))` and leave `scale_tril` as-is (because it is stored in `unbroadcasted_scale_tril`).

**Comment by fatvlady** [comment]
@fehiepsi Is there any particular reason why broadcasted `self.scale_tril` and company are stored inside the distribution object? I mean, if we really want to store it so user can see the property, why it should be broadcasted at all? I see usages only within `expand` method, and it only expands covariances to match new batch shape.

**Comment by fehiepsi** [comment]
I am not sure but I think it comes from the original design decision: all parameters should have correct batch_shape.

**Comment by ezyang** [comment]
cc @ngimel @albanD

**Comment by ezyang** [comment]
It seems to me that broadcasting tensors together and then slicing them after is a bad idea and you should avoid doing it. If there are ways we can make it easier to do the right thing in the API, those might be worth adding.

**Comment by fritzo** [comment]
Would anyone be able to help me create a `torch.broadcast_shape()` helper compatible with Pyro's, which could then be used to broadcast without slicing? I would be happy to write the unit tests, docstring and C++ implementation, but I have not touched PyTorch C++ internals in a long time and I would need a lot of help with function registration and which-files-to-touch. Alternatively I could review or collaborate on a PR.

To give you an idea of how `broadcast_shape()` works, here are some example tests:

<details>

```py
def naive_implementation(*shapes):
    result = [1] * max(len(shape) for shape in shapes)
    for shape in shapes:
        for i in range(-1, -1 - len(shape), -1):
            size = shape[i]
            if size != 1:
                result[i] = size
    return torch.Size(result)

# (good_inputs, expected_output)
GOOD_EXAMPLES = [
    ([(1, 2), (3, 1)], (3, 2)),
    ([(3, 2), (3, 1)], (3, 2)),
    ([(3, 1), (3, 2)], (3, 2)),
    ([(2, 1, 1), (1, 3, 1), (1, 1, 4)], (2, 3, 4)),
]

# bad_inputs
BAD_EXAMPLES = [
  [(2,), (3,)],
]

class TestBroadcastShape(TestCase):
    def test_ok(self):
        for inputs, output in GOOD_EXAMPLES:
            inputs = map(torch.Size, inputs)
            output = torch.Size(output)
            assert torch.broadcast_shape(*inputs) == naive_implementation(*inputs)

    def test_error(self):
        for inputs in BAD_EXAMPLES:
            inputs = map(torch.Size, inputs)
            with self.assertRaises(ValueError):
                torch.broadcast_shape(*inputs)
```

</details>

**Comment by albanD** [comment]
@fatvlady do we only need to update the `MultivariateNormal` implementation to avoid doing this expand that is not useful? 
Also I am quite confused by the `MultivariateNormal.expand()` function that doesn't seem to be used in the file.
Is there any downside of this?
If it's only the extra call to expand, would the `broadcast_shape()` proposed above solve this?

@fritzo you can just add that function in `torch/functional.py` and it will be imported automatically in the `torch.` namespace. Would that work?

**Comment by fatvlady** [comment]
@albanD Original proposal was just to fix unnecessary `loc` expand in MultivariateNormal, but it seems that it requires some changes on C++ side so that broadcast_shapes can be seamlessly used within JIT tracer.

**Comment by fritzo** [comment]
@albanD @fatvlady are you ok with the proposed fix in #43935?

@albanD the `.expand()` method is a user-facing helper that is heavily used in Pyro to implement plates in graphical models.

**Comment by albanD** [comment]
Thanks for the detail on the expand!

The PR looks like the right way for me yes.

**Comment by fatvlady** [comment]
Hi! I am trying to add `broadcast_shapes` function to C++ but I do not see in documentation how to specify `torch.Size` return type in `native_functions.yaml`. Is it currently possible or it is better to find another workaround like having optional `head` argument in `broadcast_tensors`?

**Comment by fritzo** [comment]
@fatvlady (sorry for abandoning #43935). I wonder if we could implement a fake version that inputs and outputs fake tensors with zero stride. Would the following be compatible with the jit?

<details>

```py
def broadcast_shapes(*shapes):
    scalar = torch.zeros((), device="cpu")
    tensors = [scalar.expand(shape) for shape in shapes]
    tensors = torch.broadcast_all(*tensors)
    return tensors[0].shape
```

</details>

**Comment by fatvlady** [comment]
@fritzo I think that the latter solution, despite of being somehow workaroundish, works with tracing and is definitely easier than fixing all the occurrences of `broadcast_tensors` (especially CUDA op fuser - something I would rather not to touch with my current codebase knowledge).
